File: Python/Euler023.py
#!/usr/bin/env python3
from itertools import takewhile, combinations_with_replacement
import math
import sys
if sys.version.startswith('2'):
    range = xrange


def main():
    target = 28123
    abundants = list(takewhile(lambda x: x <= target, abundant_numbers()))
    abundant_sums = {sum(pair)
                     for pair in combinations_with_replacement(abundants, 2)
                     if sum(pair) <= target}
    sum_of_abundant_sums = sum(abundant_sums)

    sum_of_all = (target + 1) // 2 * target

    sum_of_non_abundant_sums = sum_of_all - sum_of_abundant_sums

    print(sum_of_non_abundant_sums)


def abundant_numbers():
    n = 1
    while True:
        if is_abundant(n):
            yield n
        n += 1


def is_abundant(n):
    return sum(get_factors(n)) > n


def get_factors(n):
    limit = int(math.floor(math.sqrt(n))) + 1
    yield 1
    for i in range(2, limit):
        div, mod = divmod(n, i)
        if mod == 0:
            yield div
            if div * div != n:
                yield n // div


if __name__ == '__main__':
    main()

File: Node/Euler023.js
#!/usr/bin/env node
const generators = require("./generators.js");
const combinatorics = require("./combinatorics.js");

function main() {
  const target = 28123;

  var abundants = generators.takeWhile(abundantNumbers(), (x) => x <= target);
  abundants = Array.from(abundants);
  var abundantSums = combinatorics.combinationsWithReplacement(abundants, 2);
  abundantSums = generators.map(abundantSums, (pair) => pair.reduce((a, b) => a + b));
  abundantSums = generators.filter(abundantSums, (x) => x <= target);
  abundantSums = new Set(abundantSums);

  var sumOfAbundantSums = generators.reduce(abundantSums, 0, (a, b) => a + b);
  var sumOfAll = (target + 1) / 2 * target;
  var sumOfNonAbundants = sumOfAll - sumOfAbundantSums;
  console.log(sumOfNonAbundants);
}

function* abundantNumbers() {
  var n = 1;
  while (true) {
    if (isAbundant(n)) {
      yield n;
    }
    n++;
  }
}

function isAbundant(n) {
  return generators.reduce(getFactors(n), 0, (a, b) => a + b) > n;
}

function* getFactors(n) {
  var limit = Math.floor(Math.sqrt(n)) + 1;
  yield 1;
  for (let i = 2; i < limit; i++) {
    if (n % i == 0) {
      yield i;
      if (i * i != n)
        yield n / i;
    }
  }
}

if (require.main == module) {
  main()
}
